internal/provider/resource_gitlab_branch.go

```diff
@@ -0,0 +1,254 @@
+			Computed:    true,
+		},
+		"committed_date": {
+			Description: "The date at which the commit was pushed (format: yyyy-MM-ddTHH:mm:ssZ).",
+			Type:        schema.TypeString,
+			Computed:    true,
+		},
+		"committer_email": {
+			Description: "The email of the user that committed.",
+			Type:        schema.TypeString,
+			Computed:    true,
+		},
+		"committer_name": {
+			Description: "The name of the user that committed.",
+			Type:        schema.TypeString,
+			Computed:    true,
+		},
+		"short_id": {
+			Description: "The short id assigned to the commit by Gitlab.",
+			Type:        schema.TypeString,
+			Computed:    true,
+		},
+		"title": {
+			Description: "The title of the commit",
+			Type:        schema.TypeString,
+			Computed:    true,
+		},
+		"message": {
+			Description: "The commit message",
+			Type:        schema.TypeString,
+			Computed:    true,
+		},
+		"parent_ids": {
+			Description: "The id of the parents of the commit",
+			Type:        schema.TypeSet,
+			Computed:    true,
+			Elem:        &schema.Schema{Type: schema.TypeString},
+			Set:         schema.HashString,
+		},
+	},
+}
+
+func resourceGitlabBranchCreate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client := meta.(*gitlab.Client)
+	name := d.Get("name").(string)
+	project := d.Get("project").(string)
+	ref := d.Get("ref").(string)
+	branchOptions := &gitlab.CreateBranchOptions{
+		Branch: &name, Ref: &ref,
+	}
+
+	log.Printf("[DEBUG] create gitlab branch %s for project %s with ref %s", name, project, ref)
+	branch, resp, err := client.Branches.CreateBranch(project, branchOptions, gitlab.WithContext(ctx))
+	if err != nil {
+		log.Printf("[DEBUG] failed to create gitlab branch %v response %v", branch, resp)
+		return diag.FromErr(err)
+	}
+	d.SetId(buildTwoPartID(&project, &name))
+	return resourceGitlabBranchRead(ctx, d, meta)
+}
+
+func resourceGitlabBranchRead(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client := meta.(*gitlab.Client)
+	project, name, err := parseTwoPartID(d.Id())
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	log.Printf("[DEBUG] read gitlab branch %s", name)
+	branch, resp, err := client.Branches.GetBranch(project, name, gitlab.WithContext(ctx))
+	if err != nil {
+		if is404(err) {
+			log.Printf("[DEBUG] recieved 404 for gitlab branch %s, removing from state", name)
+			d.SetId("")
+			return diag.FromErr(err)
+		}
+		log.Printf("[DEBUG] failed to read gitlab branch %s response %v", name, resp)
+		return diag.FromErr(err)
+	}
+	ref := d.Get("ref").(string)
+	if ref == "" {
+		c := &gitlab.GetCommitRefsOptions{
+			Type: gitlab.String("branch"),
+		}
+		commitRefs, _, err := client.Commits.GetCommitRefs(project, branch.Commit.ID, c, gitlab.WithContext(ctx))
+		if err != nil {
+			return diag.FromErr(err)
+		}
+		for _, br := range commitRefs {
+			if br.Name != branch.Name {
+				ref = br.Name
+				break
+			}
+		}
+	}
+	d.Set("ref", ref)
```

Comment:
Isn't `ref` only known during creation time, since it is the commit or branch that you use to create this branch? I think you should move `d.Set("ref", ref)` to the CREATE function, and delete this code in the READ function. Note you will also need to set `ref` as an ignored attribute with [ImportStateVerifyIgnore](https://www.terraform.io/plugin/sdkv2/testing/acceptance-tests/teststep#importstateverifyignore) in your `Test ImportState` test step.

Reply:
I will try out implementing this as sounds slightly more watertight than the current solution thanks